internal/internal_eager_activity.go

```diff
@@ -76,75 +74,62 @@ func (e *eagerActivityExecutor) applyToRequest(
 				attrs.RequestEagerExecution = false
 			} else {
 				// If it has been requested, attempt to reserve one pending
-				attrs.RequestEagerExecution = e.reserveOnePendingSlot()
-				if attrs.RequestEagerExecution {
-					amountActivitySlotsReserved++
+				maybePermit := e.reserveOnePendingSlot()
+				if maybePermit != nil {
+					reservedPermits = append(reservedPermits, maybePermit)
+					attrs.RequestEagerExecution = true
 					eagerRequestsThisTask++
+				} else {
+					attrs.RequestEagerExecution = false
 				}
 			}
 		}
 	}
-	return
+	return reservedPermits
 }
 
-func (e *eagerActivityExecutor) reserveOnePendingSlot() bool {
-	// Lock during count checks. Nothing in here blocks including the channel
-	// receive to serve a slot.
-	e.countLock.Lock()
-	defer e.countLock.Unlock()
-	// Confirm that, if we have a max, held count isn't already there
-	if e.maxConcurrent > 0 && e.heldSlotCount >= e.maxConcurrent {
+func (e *eagerActivityExecutor) reserveOnePendingSlot() *SlotPermit {
+	// Confirm that, if we have a max, issued count isn't already there
```

Comment:
Why was the previous logic not sufficient? It already handled the case where `maxConcurrent` could be unset.

Reply:
This is the same logic - `heldSlotCount` is gone now

Reply:
Brought this back per discussion